### PR TITLE
Support for Version history

### PR DESCRIPTION
**Pull request by squallstar**
Todo (all done)

- ~UI WIP~
- ~There's a few bugs to iron out~

---

# UI

Full flow https://share.getcloudapp.com/04uKjk4R

## List of versions

image

---

## Previewing a version

image

### CODE REVIEW

File: js/interface.js [hunk 1]
```diff
@@ -909,9 +1002,65 @@ $('#app')
...
+  .on('click', '[data-version-copy]', function (e) {
+    e.preventDefault();
+    var id = $(this).data('version-copy');
+
+    return createDataSource({
+      version: {
+        dataSourceId: currentDataSourceId,
+        id: id
+      }
+    }).then(function () {
```
Comment (tonytlwu) [inline js/interface.js]: Can we pass to this function `Copy of $dataSourceName` as the default name? Users can then change it in the prompt.

Comment (squallstar) [reply js/interface.js]: done

File: js/interface.js [hunk 2]
```diff
@@ -478,11 +478,13 @@ function browseDataSource(id) {
...
 }
 
-function createDataSource(createOptions) {
+function createDataSource(createOptions, options) {
   createOptions = createOptions || {};
+  options = options || {};
 
   return Fliplet.Modal.prompt({
     title: 'Enter the name of your new Data Source',
+    value: _.get(options, 'value', '')
```
Comment (tonytlwu) [inline js/interface.js]: Semantically, the option provided is for the data source name, not for data source value.

```suggestion
    value: _.get(options, 'name', '')
```

File: templates/versions.interface.hbs [hunk 3]
```diff
@@ -22,11 +22,17 @@
...
-              <li><a href="#" data-version-preview="{{id}}">Preview data</a></li>
-              <li><a href="#" data-version-restore="{{id}}">Restore</a></li>
-              <li><a href="#" data-version-copy="{{id}}">Copy to new data source</a></li>
-            </ul>
+            <div class="btn-group" role="group">
+              <button type="button" class="btn btn-default dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                Actions
+                <span class="caret"></span>
+              </button>
+              <ul class="dropdown-menu">
```
Comment (tonytlwu) [inline templates/versions.interface.hbs]: With the example on https://getbootstrap.com/docs/3.4/javascript/#dropdowns, if you add the following styles to the `UL`, it'll align the dropdown to the right edge. You won't need to set the actions column with a fixed width. This assumes `.btn-group` has `position: relative`.

```css
    left: auto;
    right: 0;
```

File: js/interface.js (hunk 2 above)
Comment (squallstar) [reply js/interface.js]: done

File: templates/versions.interface.hbs (hunk 3 above)
Comment (squallstar) [reply templates/versions.interface.hbs]: thanks, works 
image

**Pull request merged**